tools/filamesh/src/MeshWriter.cpp

```diff
@@ -34,28 +34,79 @@ void write(ostream& out, const T* data, uint32_t count) {
     out.write((const char*) data, sizeof(T) * count);
 }
 
-void MeshWriter::serialize(ostream& out, const Mesh& mesh) {
+template<typename T>
+size_t write(unsigned char* out, const vector<T>& data) {
+    memcpy(out, data.data(), data.size() * sizeof(T));
+    return data.size() * sizeof(T);
+}
 
-    const uint32_t maxint = std::numeric_limits<uint32_t>::max();
+void MeshWriter::serialize(ostream& out, const Mesh& mesh) {
+    const uint32_t maxint = numeric_limits<uint32_t>::max();
     const bool hasIndex16 = mesh.vertexCount < maxint;
     const bool hasUV1 = !mesh.uv1.empty();
+    const size_t vertexSize = sizeof(Vertex) + (hasUV1 ? sizeof(ushort2) : 0);
+    if ((mFlags & INTERLEAVED) && hasUV1) {
+        cerr << "Interleaved vertices can only have 1 UV set." << endl;
+        exit(1);
+    }
 
     // Compute the overall bounding box.
     Box aabb = mesh.parts.at(0).aabb;
     for (size_t i = 1; i < mesh.parts.size(); i++) {
         aabb.unionSelf(mesh.parts.at(i).aabb);
     }
 
+    // Perform compression of vertex data if it has been requested.
+    vector<unsigned char> compressedVertices;
+    if (mFlags & COMPRESSION_MESHOPT) {
+        compressedVertices.resize(meshopt_encodeVertexBufferBound(mesh.vertexCount, vertexSize));
```

Comment:
Should we run the other optimization passes offered by meshoptimizer?

Reply:
Yes, I'll throw in the vertex cache optimizer, although assimp is actually already doing that for us theoretically (`aiProcess_ImproveCacheLocality`)

Reply:
Yeah but I think meshoptimizer's readme mentions its own optimization algorithms might help compression (including quantization)

Reply:
Pushed a commit that adds the other optimizations, seems to work.